bee-network/src/peer/list.rs

```diff
@@ -0,0 +1,469 @@
+        }
+
+        if self.is(peer_id, |_, state| state.is_connected()).unwrap_or(false) {
+            return Err(Error::PeerIsConnected(*peer_id));
+        }
+
+        if peer_info.relation.is_unknown() && self.count_if(|info, _| info.relation.is_unknown()) >= max_unknown_peers()
+        {
+            return Err(Error::ExceedsUnknownPeerLimit(max_unknown_peers()));
+        }
+
+        // At this point we know that the candidate is/has:
+        // * not local id
+        // * not local addr
+        // * not banned id
+        // * not banned addr
+        // * not already connected
+        // * not exceeding unknown peer limit
+
+        Ok(())
+    }
+
+    pub fn allows_dialing_peer(&self, peer_id: &PeerId) -> Result<(), Error> {
+        if peer_id == &self.local_id {
+            return Err(Error::PeerIsLocal(*peer_id));
+        }
+
+        if !self.contains(peer_id) {
+            return Err(Error::PeerNotPresent(*peer_id));
+        }
+
+        if self.banned_peers.contains(peer_id) {
+            return Err(Error::PeerIsBanned(*peer_id));
+        }
+
+        if self.is(peer_id, |_, state| state.is_connected()).unwrap_or(false) {
+            return Err(Error::PeerIsConnected(*peer_id));
+        }
+
+        let (peer_info, _) = self.peers.get(peer_id).unwrap();
+
+        if self.local_addrs.contains(&peer_info.address) {
+            return Err(Error::AddressIsLocal(peer_info.address.clone()));
+        }
+
+        if self.banned_addrs.contains(&peer_info.address) {
+            return Err(Error::AddressIsBanned(peer_info.address.clone()));
+        }
+
+        if peer_info.relation.is_unknown() && self.count_if(|info, _| info.relation.is_unknown()) >= max_unknown_peers()
+        {
+            return Err(Error::ExceedsUnknownPeerLimit(max_unknown_peers()));
+        }
+
+        // At this point we know that the candidate is/has:
+        // * not local id
+        // * present
+        // * not banned id
+        // * not already connected
+        // * not local addr
+        // * not banned addr
+        // * not exceeding unknown peer limit
+
+        Ok(())
+    }
+
+    pub fn allows_dialing_addr(&self, addr: &Multiaddr) -> Result<(), Error> {
+        if self.local_addrs.contains(addr) {
+            return Err(Error::AddressIsLocal(addr.clone()));
+        }
+
+        if self.banned_addrs.contains(addr) {
+            return Err(Error::AddressIsBanned(addr.clone()));
+        }
+
+        if let Some(peer_id) = self.find_peer_if_connected(addr) {
+            return Err(Error::PeerIsConnected(peer_id));
+        }
+
+        // At this point we know that the candidate is/has:
+        // * not local addr
+        // * not banned addr
+        // * not already connected to that addr
+
+        Ok(())
+    }
+
+    fn find_peer_if_connected(&self, addr: &Multiaddr) -> Option<PeerId> {
+        for (peer_id, _) in self.iter_if(|info, state| state.is_connected() && info.address == *addr) {
```

Comment:
isn't this the same as calling next?

Reply:
That's indeed something where I don't understand why I implemented it like this. I should just `.get` and `.filter` here.

Reply:
hmm, okay, took a deeper look at it again. I see what I did there :trollface: 

I am trying to get the `PeerId` that belongs to a certain `Multiaddr`, but I only want to include a subset (i.e. all currently connected peers) in the search.

Reply:
So what `next` do you mean?

Reply:
changed it!